src/osd/ReplicatedPG.cc

```diff
@@ -7994,9 +7989,10 @@ void ReplicatedPG::handle_watch_timeout(WatchRef watch)
   }
 
   // obc ref swallowed by repop!
-  issue_repop(repop, repop->ctx->mtime);
-  eval_repop(repop);
-  repop->put();
+  simple_repop_submit(repop);
+
+  // apply new object state.
+  ctx->obc->obs = ctx->new_obs;
```

Comment:
Hmm, I think this is cheating slightly.  It means that the watcher will appear to disappear from the perspective of a notify or list-watchers or whatever before that change is committed.  I think it would be better if we could call the do_osd_op_effects() so that the code path is closer to the normal unwatch case.